createResolversCacheMiddleware: remove dependency on core/data store

jsnajdr opened the pull request:
The `core/data` store is a legacy store that provides access to "resolution state" of other stores' resolvers, and exists only to provide backward compatibility. The supported way to access store's "resolution state" is to use its metadata selectors like `hasFinishedResolution`.

This PR removes the `core/data` dependency from the resolvers cache middleware, which is used by all stores, and select directly from related stores instead.

I'm also changing the code slightly to reduce one level of indentation, and removing unneeded optional chaining when looking at `value.status`.

**How to test:**
Sufficiently covered by existing unit tests.

jsnajdr committed:
createResolversCacheMiddleware: remove dependency on core/data store

Mamaduka reviewed (approved):
A nice refactoring and the changes look good to me ✅

P.S. It was easier to review with hidden whitespaces - https://github.com/WordPress/gutenberg/pull/54733/files?diff=unified&w=1.

tyxla reviewed (approved):
Great catch! 💯 

LGTM 🚀

tyxla commented:
Hmm, it appears that the failing tests might be related 🤔

jsnajdr commented:
There were e2e failures that prevent me from merging. I rebased the PR onto the latest `trunk`, let's see if CI passes now. If it doesn't, I'll start to investigate whether I introduced some bug.

tyxla commented on packages/data/src/resolvers-cache-middleware.js:
```diff
@@ -1,53 +1,41 @@
+				// If the value is "finished" or "error" it means this resolver has finished its resolution which means we need
+				// to invalidate it, if it's true it means it's inflight and the invalidation is not necessary.
+				if (
+					( value.status !== 'finished' &&
```
Judging by the (previously) failing tests, it looks like those optional chaining instances were necessary.

jsnajdr replied:
Yes, thanks for the hint! The `value` can indeed be `undefined`, although our code always sets it to a valid object value. It's a bug in `EquivalentKeyMap` where `map.delete` merely sets a map entry value to `undefined`:

https://github.com/aduth/equivalent-key-map/blob/d4115f9fbd5d1bf27de58be8d9c0e50eb5ede452/src/equivalent-key-map.js#L204

and then the `map.forEach` method iterates also through these deleted values, not only through valid ones.

jsnajdr committed:
Work around bug in EquivalentKeyMap delete method

tyxla reviewed (approved):
Thanks for the fix, looks good and tests well 👍 🚀
